### PR TITLE
Assumptions stuff

### PR DESCRIPTION
**Pull request by o-**
(see commits)

### CODE REVIEW

File: rir/src/compiler/opt/delay_instr.cpp [hunk 1]
```diff
@@ -15,8 +15,12 @@ void DelayInstr::apply(RirCompiler&, Closure* function, LogStream&) const {
             auto i = *ip;
             auto next = ip + 1;
 
-            if (!i->hasEnv() && !i->hasEffect() && !Phi::Cast(i) &&
-                !i->branchOrExit()) {
+            if (LdFun::Cast(i) ||
+                // For LdFun we allow the instruction to be delayed, since
+                // probably nobody cares about when exactly looking up a
```
Comment (charig) [inline rir/src/compiler/opt/delay_instr.cpp]: Mmmm, just wondering, if this is the case, maybe we can just declare them effectfulness and move the comment there?

File: rir/src/compiler/opt/pass_definitions.h [hunk 2]
```diff
@@ -105,6 +105,12 @@ class PASS(CleanupCheckpoints, "Cleanup unused checkpoints");
  */
 class PASS(CleanupFramestate, "Cleanup framestates unused by checkpoints");
 
+/*
+ * Trying to group assumptions, by pushing them up. This well lead to fewer
```
Comment (charig) [inline rir/src/compiler/opt/pass_definitions.h:109]: There is a typo: `well -> will`. But now that I already write this, I would rephrase:
`Tries to group assumptions to reduce the number of checkpoints used overall`

File: rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp [hunk 3]
```diff
@@ -297,15 +298,18 @@ bool Rir2Pir::compileBC(const BC& bc, Opcode* pos, Opcode* nextPos,
             }
 
             std::string name = "";
-            if (auto ldfun = LdFun::Cast(callee))
+            if (ldfun)
```
Comment (charig) [inline rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:301]: why not just put the `auto ldfun = LdFun::Cast(callee);` here, in line 300?

File: rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp [hunk 4]
```diff
@@ -297,15 +298,18 @@ bool Rir2Pir::compileBC(const BC& bc, Opcode* pos, Opcode* nextPos,
...
             Assumptions asmpt(Assumption::CorrectOrderOfArguments);
             asmpt.set(Assumption::CorrectNumberOfArguments);
             compiler.compileClosure(
                 monomorphic, name, asmpt,
                 [&](Closure* f) {
                     Value* expected = insert(new LdConst(monomorphic));
-                    Value* t = insert(new Identical(callee, expected));
+                    Value* given = callee;
+                    if (ldfun)
+                        given = insert(new LdVar(ldfun->varName, ldfun->env()));
```
Comment (charig) [inline rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:313]: I think there are two scenarios here.
1. There is a binding in the environment.
2. There is not.
For case number 2 I do not see how this helps. Even worst, it adds an instruction. For case number 1 well, if I guess you are looking for the case we can prove the binding is a constant and thus we can remove the whole speculation.
In case I am right maybe it would better to put this on pir instead of rir...

File: rir/src/compiler/opt/assumptions.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    Visitor::run(function->entry, [&](BB* bb) {
+        auto ip = bb->begin();
+        while (ip != bb->end()) {
+            auto next = ip + 1;
+            auto instr = *ip;
+
+            if (auto assume = Assume::Cast(instr)) {
+                auto cp = assume->checkpoint();
+                auto bb = cp->bb();
+                // We are trying to group multiple assumes into the same
```
Comment (charig) [inline rir/src/compiler/opt/assumptions.cpp:27]: I was a little confused by the comment. `moving up linear sequence of bbs` made me think we are moving bbs.
I would say: `We are trying to group multiple assumes into the same checkpoint by finding for each assume the topmost compatible checkpoint.`
    
Then, again, just brainstorming...I have the impression this is becoming a kind of pattern we are facing many times. 
Maybe we can add to cfg a function `findTopmostComplyingWith(Instruction inst, lambda condition)` and iterate it till we reach a general version. I mean the code here would be much easier to read if we replace the whole while with

**Review by charig — APPROVED** [review]

**Comment by o-** [comment]
@charig so maybe to explain better my idea in the first commit.

Instead of this:
```
    f = ldfun("f")
    c = checkpoint()  -> BB1 | BB2
BB2:
    deopt(stack=[f])
BB1:
    assume (f==someFunction, c)
```

I want this:
```
    c = checkpoint()  -> BB1 | BB2
BB2:
    f1 = ldfun("f")
    deopt(stack=[f1])
BB1:
    f2 = ldvar("f")
    assume (f2==someFunction, c)
```

because f2 has no sideeffect the code is better for optimizations.

(I should probably add that diagram to the code)

I don't think that we can mark ldfun as having no sideeffects in general, since that would certainly break some optimiztions. I think what would be better is to make it even more explicit, ie. just allow ldfun to move over checkpoint, because that is always safe to do. The reason is that we do not allow sideeffects between assume and checkpoint, so the ldfun sideeffect in the example above still happens at the correct time.

**Comment by charig** [comment]
> @charig so maybe to explain better my idea in the first commit.
> 
> Instead of this:
> 
> ```
>     f = ldfun("f")
>     c = checkpoint()  -> BB1 | BB2
> BB2:
>     deopt(stack=[f])
> BB1:
>     assume (f==someFunction, c)
> ```
> I want this:
> 
> ```
>     c = checkpoint()  -> BB1 | BB2
> BB2:
>     f1 = ldfun("f")
>     deopt(stack=[f1])
> BB1:
>     f2 = ldvar("f")
>     assume (f2==someFunction, c)
> ```
> because f2 has no sideeffect the code is better for optimizations.
> 
> (I should probably add that diagram to the code)
> 
> I don't think that we can mark ldfun as having no sideeffects in general, since that would certainly break some optimiztions. I think what would be better is to make it even more explicit, ie. just allow ldfun to move over checkpoint, because that is always safe to do. The reason is that we do not allow sideeffects between assume and checkpoint, so the ldfun sideeffect in the example above still happens at the correct time.

Ok. But then in your code there is a hidden assumption: that the callee `ldFun f` is not used anywhere else and thus, a later delay pass will move it to the deopt branch. If this is not true then we may end up with worst code instead of better. I do not know if ldFun could be used more than once but I assume so.

**Comment by o-** [comment]
> But then in your code there is a hidden assumption: that the calle ldFun f is not used anywhere else and thus, a later delay pass will move it to the deopt branch. If this is not true then we may end up with worst code instead of better. 

that is true. it's very hard to know in advance if an optimization is a net win. e.g. maybe there is a second use, but the second use can be resolved by the same trick...
